File: csharp/patterns/src/PrototypePatterns/Creational/ObjectPool/ObjectPoolPattern.cs
namespace PrototypePatterns.Creational.ObjectPool;

public sealed class Worker(int workerId)
{
    public int WorkerId { get; } = workerId;
    public bool Busy { get; set; }

    public void Reset() => Busy = false;
}

public sealed class WorkerPool
{
    private readonly Queue<Worker> _available = new();
    private readonly HashSet<int> _inUse = [];

    public WorkerPool(int size)
    {
        for (var index = 0; index < size; index++)
        {
            _available.Enqueue(new Worker(index));
        }
    }

    public Worker Acquire()
    {
        if (_available.Count == 0)
        {
            throw new InvalidOperationException("pool exhausted");
        }

        var worker = _available.Dequeue();
        worker.Busy = true;
        _inUse.Add(worker.WorkerId);
        return worker;
    }

    public void Release(Worker worker)
    {
        if (!_inUse.Remove(worker.WorkerId))
        {
            throw new InvalidOperationException("worker was not checked out");
        }

        worker.Reset();
        _available.Enqueue(worker);
    }
}

public static class ObjectPoolPattern
{
    public static bool Demo()
    {
        var pool = new WorkerPool(2);
        var first = pool.Acquire();
        pool.Acquire();
        pool.Release(first);
        var again = pool.Acquire();
        return first.WorkerId == again.WorkerId;
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/creational/objectpool/ObjectPoolPattern.java
package org.prototypepatterns.creational.objectpool;

import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

public final class ObjectPoolPattern {
    private ObjectPoolPattern() {
    }

    public static final class Worker {
        public final int workerId;
        public boolean busy;

        public Worker(int workerId) {
            this.workerId = workerId;
        }

        public void reset() {
            busy = false;
        }
    }

    public static final class WorkerPool {
        private final Queue<Worker> available = new ArrayDeque<>();
        private final Set<Integer> inUse = new HashSet<>();

        public WorkerPool(int size) {
            for (int i = 0; i < size; i++) {
                available.add(new Worker(i));
            }
        }

        public Worker acquire() {
            Worker worker = available.poll();
            if (worker == null) {
                throw new IllegalStateException("pool exhausted");
            }
            worker.busy = true;
            inUse.add(worker.workerId);
            return worker;
        }

        public void release(Worker worker) {
            if (!inUse.remove(worker.workerId)) {
                throw new IllegalStateException("worker was not checked out");
            }
            worker.reset();
            available.add(worker);
        }
    }

    public static boolean demo() {
        WorkerPool pool = new WorkerPool(2);
        Worker first = pool.acquire();
        pool.acquire();
        pool.release(first);
        Worker again = pool.acquire();
        return first.workerId == again.workerId;
    }
}
